openquake/hazardlib/gsim/kotha_2019.py

```diff
@@ -0,0 +1,375 @@
+# along with OpenQuake. If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Module exports :class:`KothaEtAl2019`,
+               :class:`KothaEtAl2019SERA`
+"""
+import os
+import h5py
+import numpy as np
+from scipy.constants import g
+from scipy.interpolate import interp1d
+from openquake.hazardlib.gsim.base import GMPE, CoeffsTable
+from openquake.hazardlib import const
+from openquake.hazardlib.imt import PGA, PGV, SA, from_string
+
+
+BASE_PATH = os.path.join(os.path.dirname(__file__), "kotha_2019_tables")
+
+
+class KothaEtAl2019(GMPE):
+    """
+    Implements the first complete version of the newly derived GMPE
+    for Shallow Crustal regions using the Engineering Strong Motion Flatfile.
+
+    (Working Title)
+    Kotha, S. R., Weatherill, G., Bindi, D., Cotton F. (2019) A Revised
+    Ground Motion Prediction Equation for Shallow Crustal Earthquakes in
+    Europe and the Middle-East
+
+    The GMPE is desiged for calibration of the stress parameter term
+    (a multiple of the fault-to-fault variability, tau_f) an attenuation
+    scaling term (c3) and a statistical uncertainty term (sigma_mu). The
+    statistical uncertainty is a scalar factor dependent on period, magnitude
+    and distance. These are read in from hdf5 upon instantiation and
+    interpolated to the necessary values.
+
+    In the core form of the GMPE no site term is included. This will be
+    added in the subclasses.
+    """
+    experimental = True
+
+    #: Supported tectonic region type is 'active shallow crust'
+    DEFINED_FOR_TECTONIC_REGION_TYPE = const.TRT.ACTIVE_SHALLOW_CRUST
+
+    #: Set of :mod:`intensity measure types <openquake.hazardlib.imt>`
+    #: this GSIM can calculate. A set should contain classes from module
+    #: :mod:`openquake.hazardlib.imt`.
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set([
+        PGA,
+        PGV,
+        SA
+    ])
+
+    #: Supported intensity measure component is the geometric mean of two
+    #: horizontal components
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.RotD50
+
+    #: Supported standard deviation types are inter-event, intra-event
+    #: and total
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set([
+        const.StdDev.TOTAL,
+        const.StdDev.INTER_EVENT,
+        const.StdDev.INTRA_EVENT
+    ])
+
+    #: Required site parameter is not set
+    REQUIRES_SITES_PARAMETERS = set()
+
+    #: Required rupture parameters are magnitude and hypocentral depth
+    REQUIRES_RUPTURE_PARAMETERS = set(('mag', 'hypo_depth'))
+
+    #: Required distance measure is Rjb (eq. 1).
+    REQUIRES_DISTANCES = set(('rjb', ))
+
+    def __init__(self, stress_epsilon=0.0, c3=None, sigma_mu_epsilon=0.0):
+        """
+        Instantiate setting the stress_epsilon, sigma_mu_epsilon and c3 terms
+        """
+        super().__init__()
+        self.stress_epsilon = stress_epsilon
+        if isinstance(c3, dict):
+            # Inputing c3 as a dictionary sorted by the string representation
```

Comment:
I can add a check to make sure the SA coefficients are sorted, and sort them if not. However, @micheles should this actually be done in the CoeffsTable object itself rather than the GMPE to catch any case that someone creates the CoeffsTable from a dictionary that is not in order?

Reply:
But why the dictionary should be sorted? I have the impression everything will work even if it is sorted in a different order. BTW, since Python 3.6 the order of dictionaries is fixed, so if the CoeffsTable comes from a table in string form (as usual) the order of the fields will be preserved, i.e. they keys will be in the same order as the columns.

Reply:
The issue was mostly to address the case that someone instantiates the GMPE using a dictionary in which the SA() keys are not in order of ascending period on input (for example in the xml GMPE logic tree file).